File: plugin/runtime/src/main/java/com/jaynewstrom/json/runtime/internal/ListSerializer.java
package com.jaynewstrom.json.runtime.internal;

import com.fasterxml.jackson.core.JsonGenerator;
import com.jaynewstrom.json.runtime.JsonSerializer;
import com.jaynewstrom.json.runtime.JsonSerializerFactory;

import java.io.IOException;
import java.util.List;

public final class ListSerializer<T> implements JsonSerializer<List<T>> {
    private final JsonSerializer<T> elementSerializer;

    public ListSerializer(JsonSerializer<T> elementSerializer) {
        this.elementSerializer = elementSerializer;
    }

    @Override public void serialize(List<T> elements, JsonGenerator jg, JsonSerializerFactory serializerFactory) throws IOException {
        jg.writeStartArray();
        for (int i = 0, size = elements.size(); i < size; i++) {
            elementSerializer.serialize(elements.get(i), jg, serializerFactory);
        }
        jg.writeEndArray();
    }
}

File: plugin/runtime/src/main/java/com/jaynewstrom/jsonDelight/runtime/internal/ListSerializer.kt
package com.jaynewstrom.jsonDelight.runtime.internal

import com.fasterxml.jackson.core.JsonGenerator
import com.jaynewstrom.jsonDelight.runtime.JsonSerializer
import com.jaynewstrom.jsonDelight.runtime.JsonSerializerFactory

import java.io.IOException

class ListSerializer<T>(private val elementSerializer: JsonSerializer<T>) : JsonSerializer<List<T>> {
    @Throws(IOException::class)
    override fun serialize(value: List<T>, jg: JsonGenerator, serializerFactory: JsonSerializerFactory) {
        jg.writeStartArray()
        var i = 0
        val size = value.size
        while (i < size) {
            elementSerializer.serialize(value[i], jg, serializerFactory)
            i++
        }
        jg.writeEndArray()
    }
}
